### PR TITLE
Add caching support in the local proxy server

### PR DESCRIPTION
**Pull request by tvoran**
(Continuing #52)

Adds caching support to the local proxy server for the Vault lambda extension. To use, set the VAULT_DEFAULT_CACHE_TTL environment variable to a time.Duration value (15m, 1h). Requests with the header "X-Vault-Cache-Control: cache" will be returned from the cache on a hit, and responses will be stored in the cache if not. The cache can be invalidated by setting the header to "recache".

Also if the VAULT_DEFAULT_CACHE_ENABLE environment variable is set to `true`, all requests will behave as though "X-Vault-Cache-Control: cache" was set.

Changes from the original PR:

- [x] Use go-cache for cache management
- [x] Use a hash of the full request as the cache key
- [x] Rename VAULT_CACHE_TTL -> VAULT_DEFAULT_CACHE_TTL
- [x] Use headers instead of query parameters for cache control
- [x] Option to always have caching enabled

### CODE REVIEW

File: internal/proxy/cache.go
Comment (tomhjp) [inline internal/proxy/cache.go:69]: Is it possible that the body also needs closing when there is an error above? And should we also log the returned error from `Close` if it's non-nil?

File: internal/proxy/proxy.go
Comment (tomhjp) [inline internal/proxy/proxy.go:60]: I think we should probably lock an exclusive mutex, and immediately defer the unlock, before we `Get` from the cache. If two requests come in simultaneously, the last one should wait until the first finishes. We could make the mutex specific to the particular key we're retrieving, but that seems likely to be overkill in complexity compared to the typical scale at which a single function will be making Vault requests.

**Review by tomhjp — COMMENTED** [review]
Looking good! Some questions to discuss around error handling and concurrency, but mostly nits at this stage. I haven't reviewed the tests yet, but will review on the next round.

File: README.md
Comment (tvoran) [reply README.md]: Wrapped in b260178c9678c4ab7400662fb053240a055ac2db

File: internal/proxy/cache.go
Comment (tvoran) [reply internal/proxy/cache.go]: Moved in c1dd09aa3922f8560490e0b399478ceb783842a4

File: internal/proxy/cache.go
Comment (tvoran) [reply internal/proxy/cache.go]: Yeah, I like that better. Changed in c1dd09aa3922f8560490e0b399478ceb783842a4.

File: internal/proxy/proxy.go
Comment (tvoran) [reply internal/proxy/proxy.go:15]: I think that makes sense, except for X-Vault-Cache-Control since that's specific to VLE for now.

**Comment by tvoran** [comment]
@JinlianWang Yes that was added in a0256b1.

File: internal/proxy/cache.go
Comment (tvoran) [reply internal/proxy/cache.go:193]: Yeah, I think I agree on this, and good point about the 503s. Will take a look.

File: internal/proxy/cache.go
Comment (tvoran) [reply internal/proxy/cache.go]: I have been wondering this too, and I suspect you're right that other methods should be ok to be cached. It'll just need a little more testing and investigation.

File: internal/proxy/proxy.go
Comment (tomhjp) [reply internal/proxy/proxy.go]: Yeah agreed 👍

File: internal/proxy/proxy.go
Comment (tomhjp) [reply internal/proxy/proxy.go:62]: Yeah the underlying cache access looks safe. My thinking was more at the application behaviour level. If we don't lock between a cache miss and the underlying request being made, then we could have multiple in-flight Vault requests for the same secret simultaneously. And that could also result in inconsistent results from the cache. e.g. I think currently [this](https://mermaid-js.github.io/mermaid-live-editor/edit#pako:eNp9kU-LwjAQxb9KmLNFdNVDDkLriteFXfaUy5hMtWASTROwFL-7o1S07tLb8H7vvfyZFrQ3BBJqOiVymj4r3AW0yqGOPoj8MRTK5dly-RX8uZFis_4RZjsuvVeu-F--S4x-MR2iFPkgLfr0rmbPXsapppBNph_PaMdWqPf0BzPvhd4b8w7O5ouhxhfMvBcafPbL8WwWZfBW6FstjMBSsFgZ_vNWOSEUxD1ZUiB5NFTeLqpAuQtb09FgpLWpeAMgSzzUNAJM0X83ToOMIdHD1O2tc12uGiidPA) could happen, but we want [this](https://mermaid-js.github.io/mermaid-live-editor/edit/#pako:eNqFkMEKwjAQRH8l7LlF1FsOhVbFq6B4ymVNtrbQJJomoEj_3VQqRax4W-bNg2QeIK0i4NDSNZCRtK7x7FALg9Jbx_L3UQiTp1m2c_Z252y7OTB1mpXWClNMx68ooiOGxnOWf9JXmo5mxKEll84Xy1Ed2AplRV848gkppsWYstJZzWTv_3nnTw8S0OQ01iqO9BCGMQG-Ik0CeDwVlf1PBAjTxWq4KPS0UXWcDHiJTUsJYPB2fzcSuHeB3qVh6KHVPQHhd4Nj) to happen instead. The diagram has two actors, but they could just represent different threads within the same process.

File: internal/proxy/cache.go
Comment (calvn) [inline internal/proxy/cache.go:31]: The [standard cache-control headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) use a slightly different directive for these two behavior. `no-cache` is equivalent to `recache` and `no-store` is equivalent to no cache. These are vault-specific cache headers though, so I'll leave it up to you on whether we want to use the same naming convention.

File: internal/proxy/cache.go
Comment (calvn) [reply internal/proxy/cache.go:193]: I agree as well. On Agent we [only cache 2xx (successful) responses](https://github.com/hashicorp/vault/blob/eefc9817a1147a877d77dc1700a29cab830b83a6/command/agent/cache/lease_cache.go#L283-L287).

File: internal/proxy/proxy.go
Comment (calvn) [inline internal/proxy/proxy.go:18]: Does this need to be addressed before this PR is merged?

File: internal/proxy/proxy.go
Comment (calvn) [reply internal/proxy/proxy.go:62]: We need a way to track inflight requests in order to avoid the situation that Tom described above, where there might be two concurrent inflight requests leading to two of the same requests (e.g. a creds generation request) to be made against the Vault server. 

A global exclusive mutex lock would make the entire caching mechanism synchronous and it'd only be able to process one request at a time. The[ approach that I took for agent](https://github.com/hashicorp/vault/pull/10705/files) to address this issue is to make a second and separate inflight cache that is clears once the first request "wins". The other requests will block (using a channel) until then. This is a way to coordinate concurrent inflight requests of the same type without having to manage locks per request or making the entire cache operate on a global lock.

File: internal/proxy/proxy.go
Comment (tvoran) [reply internal/proxy/proxy.go:18]: I don't believe so.

File: internal/proxy/cache.go
Comment (tvoran) [reply internal/proxy/cache.go]: I'm not really sure the body needs closing when there is an error above, since an error here will stop processing the request in the calling handler. That being said I added more logging and another call to r.Body.Close() out of an abundance of caution in 73bfb8c815700f7a4bbe6951eae49d7c95579244.

File: internal/proxy/cache.go
Comment (tvoran) [reply internal/proxy/cache.go]: I refactored the cache refresh portion in 6f639320ba45d499b1b491be433a4a11ff3b8d06 to only cache 2xx responses.

File: internal/proxy/proxy.go
Comment (tomhjp) [reply internal/proxy/proxy.go:63]: I think allowing multiple inflight requests is more complexity than we need. The scale demands on this proxy are much smaller than for Agent, because each instance of the proxy is dedicated to only one instance of the Lambda function at a time. Given that, I actually don't think a global mutex would be disastrous, but I'm happy to start with a mutex per cache key instead.

File: internal/proxy/cache.go
Comment (tomhjp) [reply internal/proxy/cache.go]: Yeah sorry, to be clear, I actually wasn't sure of the answers to my questions. As you say, handling those cases is showing an abundance of caution, but LGTM 👍

File: internal/proxy/proxy.go
Comment (tvoran) [reply internal/proxy/proxy.go:76]: Appreciate the discussion and the diagrams especially here. I added a mutex per cache key in 6e523bf, and then moved the lock acquisition to the top of the proxyHandler in 8c49e18.

I was thinking about a similar scenario with two requests, but one of them has "recache" set. The "recache" request would skip the cache.Get() (and the locking), but still get the cred from Vault and store it in the cache.

File: internal/proxy/proxy.go
Comment (tvoran) [reply internal/proxy/proxy.go]: Added them to Vault's api package in https://github.com/hashicorp/vault/pull/14067

File: internal/proxy/cache_test.go
Comment (tomhjp) [inline internal/proxy/cache_test.go:491]: nit: `retrieveData` shouldn't touch the request body, so it would be a slightly tighter test if this is nil or a different body.

File: internal/proxy/cache_test.go
Comment (tomhjp) [inline internal/proxy/cache_test.go:306]: Can we make the expiry and sleep time something like 10ms and 20ms respectively? The cache package's own tests use times of that order of magnitude, and looking at the implementation, I don't think it will be flaky.

**Review by tomhjp — APPROVED** [review]
Only minor comments - I think this looks pretty much ready to go 👍

File: README.md
Comment (tvoran) [reply README.md:174]: Agreed; I added mention of it being required in 083f08b73f2e1b1638521519fe1d3aac0d08969a.

File: README.md
Comment (tvoran) [reply README.md]: Good point! I reworked this section a bit in 083f08b73f2e1b1638521519fe1d3aac0d08969a to make it flow a little better with fewer surprises.

File: internal/proxy/cache.go
Comment (tvoran) [reply internal/proxy/cache.go]: Yep, `defaultOn` is definitely clearer. Changed in bd90b345c483bfb20438482fa489a3ef54380bfc.

File: internal/proxy/cache_test.go
Comment (tvoran) [reply internal/proxy/cache_test.go]: Done in bd90b345c483bfb20438482fa489a3ef54380bfc.

File: internal/proxy/cache.go
Comment (tvoran) [reply internal/proxy/cache.go]: Yep! Gone in bd90b345c483bfb20438482fa489a3ef54380bfc.

File: internal/proxy/cache_test.go
Comment (tvoran) [reply internal/proxy/cache_test.go]: Cool, changed it to 10ms and 20ms in bd90b345c483bfb20438482fa489a3ef54380bfc and it seems to be working fine.

File: internal/proxy/cache.go
Comment (tvoran) [reply internal/proxy/cache.go:30]: That's a good point, though I do think the headers as-is are clearer as to their meaning than those of the standard. However I don't like that there is overlap with the standard headers, since that could lead to confusion, so I'm happy to change them if we just want to stick with the standard.

Comment (calvn) [reply internal/proxy/cache.go:30]: That's fair, these are not strictly the same standard cache-control headers.

File: internal/proxy/cache.go
Comment (calvn) [inline internal/proxy/cache.go:163]: If we support passing in more that one caching option (with the most restrictive one winning), then we should also parse for the case we're `X-Vault-Cache-Control` is passed more than once since that's [equivalent](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2) to passing a comma-separated list

**Review by calvn — APPROVED** [review]
One small comment around header parsing, otherwise this looks great!

File: internal/proxy/cache.go
Comment (tvoran) [reply internal/proxy/cache.go]: Yep, that's a good point, added the extra parsing in 675aa462eade8a80299d379d8b89a097d19de39f.

**Pull request merged**
